model/src/main/proto/math.proto

```diff
@@ -5,17 +5,186 @@ package model;
 option java_package = "org.oppia.android.app.model";
 option java_multiple_files = true;
 
-// Structure for a fraction object.
+// Represents a fraction.
+//
+// Values of this proto can be analyzed using FractionSubject.
 message Fraction {
+  // Defines whether the fraction is negative.
   bool is_negative = 1;
-  int32 whole_number = 2;
-  int32 numerator = 3;
-  int32 denominator = 4;
+
+  // Defines the whole number component of the fraction.
+  uint32 whole_number = 2;
+
+  // Defines the numerator of the fraction.
+  uint32 numerator = 3;
+
+  // Defines the denominator of the fraction. This should never be zero.
+  uint32 denominator = 4;
 }
 
-// Structure containing a ratio object for eg - [1,2,3] for 1:2:3.
+// Represents a structured real value.
+//
+// Values of this proto can be analyzed using RealSubject.
+message Real {
+  // Defines type of real value.
+  oneof real_type {
+    // Indicates that this real value is a fraction.
+    Fraction rational = 1;
+
+    // Indicates that this real value is a decimal. Technically these can sometimes be rational, but
+    // given IEEE-754 rounding errors and the difficulty of factoring fractions, many rational
+    // decimal values need to be treated as irrational and non-factorable.
+    double irrational = 2;
+
+    // Indicates that thi sreal value is an integer (as a special case of rational values since
+    // integers are easier to work with than fraction objects). Note that this isn't the only case
+    // where the real value can be an integer. It can also be an integer double value, or a fraction
+    // with only a whole number component.
+    int32 integer = 3;
+  }
+}
+
+// Represents a ratio, e.g. 1:2:3.
 message RatioExpression {
-  // List of components in a ratio. It's expected that list should have more than
-  // 1 element.
+  // List of components in a ratio. For example, the ratio 1:2:3 will have component values
+  // [1, 2, 3]. It's expected that list should have more than 1 element.
   repeated uint32 ratio_component = 1;
 }
+
+// Represents a mathematical expression such as 1+2*7. Expressions are inherently recursive, so the
+// overall expressiveness of this structure (& math expressions as a whole) is defined based on its
+// constituent substructures.
+//
+// This structure is designed to represent both numeric and algebraic expressions.
+//
+// Values of this proto can be analyzed using MathExpressionSubject.
+message MathExpression {
+  // The index within the input text stream at which point the expression starts (it's an inclusive
+  // index). If both this and the end index are zero then no parsing information is included for
```

Comment:
```suggestion
  // index). If both this and the end index are the same then no parsing information is included for
```

It can be a value other than zero as well right?

Reply:
Good point; updated to your suggestion.